tf_encrypted/protocol/securenn/odd_tensor.py

```diff
@@ -58,170 +63,225 @@ def modulus(self):
-        return self.mul(other)
+    def __add__(self, other):
+        return self.add(other)
 
-    def __mod__(self, k: int) -> 'OddImplicitTensor':
-        return self.mod(k)
+    def __sub__(self, other):
+        return self.sub(other)
 
-    def add(self, other) -> 'OddImplicitTensor':
+    def add(self, other):
         x, y = _lift(self, other)
         bitlength = math.ceil(math.log2(self.factory.modulus))
 
-        z = x.value + y.value
+        # the below avoids redundant seed expansion; can be removed once
+        # we have a (per-device) caching mechanism in place
+        x_value = x.value
+        y_value = y.value
+
+        z = x_value + y_value
 
-        # We want to compute `pos(x) + pos(y) >= m - 1` for correction purposes
-        # which, since `m - 1 == 1` for signed integers, can be rewritten as
+        # we want to compute whether we wrapped around, ie `pos(x) + pos(y) >= m - 1`,
+        # for correction purposes which, since `m - 1 == 1` for signed integers, can be
+        # rewritten as:
         #  -> `pos(x) >= m - 1 - pos(y)`
         #  -> `m - 1 - pos(y) - 1 < pos(x)`
         #  -> `-1 - pos(y) - 1 < pos(x)`
-        wrapped_around = _lessthan_as_unsigned(-2 - y.value, x.value, bitlength)
+        #  -> `-2 - pos(y) < pos(x)`
+        wrapped_around = _lessthan_as_unsigned(-2 - y_value, x_value, bitlength)
         z += wrapped_around
 
-        return OddImplicitTensor(z, self.factory)
+        return OddDenseTensor(z, self.factory)
 
-    def sub(self, other) -> 'OddImplicitTensor':
+    def sub(self, other):
         x, y = _lift(self, other)
         bitlength = math.ceil(math.log2(self.factory.modulus))
 
-        z = x.value - y.value
+        # the below avoids redundant seed expansion; can be removed once
+        # we have a (per-device) caching mechanism in place
+        x_value = x.value
+        y_value = y.value
 
-        # We want to compute `pos(x) - pos(y) < 0` for correction purposes
-        # which can be rewritten as
+        z = x_value - y_value
+
+        # we want to compute whether we wrapped around, ie `pos(x) - pos(y) < 0`,
+        # for correction purposes which can be rewritten as
         #  -> `pos(x) < pos(y)`
-        wrapped_around = _lessthan_as_unsigned(x.value, y.value, bitlength)
+        wrapped_around = _lessthan_as_unsigned(x_value, y_value, bitlength)
         z -= wrapped_around
 
-        return OddImplicitTensor(z, self.factory)
-
-    def mul(self, other) -> 'OddImplicitTensor':
-        raise NotImplementedError()
-
-    def matmul(self, other) -> 'OddImplicitTensor':
-        raise NotImplementedError()
+        return OddDenseTensor(z, self.factory)
 
-    def im2col(self, h_filter, w_filter, padding, strides) -> 'OddImplicitTensor':
-        raise NotImplementedError()
-
-    def conv2d(self, other, strides, padding='SAME') -> 'OddImplicitTensor':
-        raise NotImplementedError()
+    def bits(self, factory=None):
+        if factory is None:
```

Comment:
Small simplification:

```
if factory is None:
  factory = self.factory

return factory.tensor(binarize(self.value))

Reply:
Noticed somewhere you mentioned direct constructor is better at compile time so maybe this isn't relevant.

Reply:
Also not even sure this function still exists 😄 !

Reply:
in this case it's actually an optimization to call the constructor directly, since it avoids the check for -1 values in the factory! works here since we know `binarize` will not return -1 values and hence no need to correct for these.